design-pattern.md

```diff
@@ -1 +1,151 @@
+<!-- _class: due -->
+
 # Design pattern
+
+I design pattern sono dei modelli di soluzione a problemi comuni di progettazione software. Esistono diversi tipi di design pattern, ma in generale si possono suddividere in 4 categorie: _creazionali, strutturali, comportamentali e architetturali_.
+
+## Pattern creazionali
+
+I design pattern creazionali sono quelli che si occupano di risolvere problemi relativi alla _creazione di oggetti_.
+
+### Factory Method
+
+Il **Factory Method** permette di creare oggetti senza specificare la loro classe concreta. Questo pattern è particolarmente utile quando si deve creare un oggetto che dipende da altri oggetti, o quando si deve creare un oggetto che dipende da parametri variabili. Infatti, il Factory Method permette di creare oggetti in modo indipendente dalle dipendenze che essi hanno.
+
+#### UML
+
+![](./immagini/factory-method.png)
+
+Nell'UML, la classe astratta `Creator`, che ha bisogno di creare un oggetto `Product`, delega la creazione ad una classe `ConcreteCreator` (che estende `Creator`). La classe `ConcreteCreator` implementa il metodo `factoryMethod()` che crea un oggetto `ConcreteProduct` (che estende `Product`).
+
+#### Esempio
+
+Supponiamo di voler creare un programma per creare dei documenti. Il programma deve essere in grado di creare documenti di diversi tipi (documenti di testo, di calcolo, pdf, ...) decisi a runtime dall'utente. Il programma non può prevedere che tipo di documento l'utente deciderà di creare, quindi il Factory Method è la soluzione migliore.
+
+```java
+public interface Document { /* ... */ }
+
+public class TextDocument implements Document { /* ... */ }
+
+public class PDFDocument implements Document { /* ... */ }
+
+public abstract class DocumentFactory {
+    public abstract Document createDocument();
+}
+
+public class TextDocumentFactory extends DocumentFactory {
+    @Override
+    public Document createDocument() {
+        return new TextDocument();
+    }
+}
+
+public class PDFDocumentFactory extends DocumentFactory {
+    @Override
+    public Document createDocument() {
+        return new PDFDocument();
+    }
+}
+
+public class Main {
+    public static void main(String[] args) {
+        DocumentFactory factory;
+        Document document;
+
+        // Creazione di un documento di testo
+        factory = new TextDocumentFactory();
+        document = factory.createDocument();
+
+        // Creazione di un documento PDF
+        factory = new PDFDocumentFactory();
+        document = factory.createDocument();
+    }
+}
+```
```

Comment:
Per collegarmi alla spiegazione che suggerivo all'inizio, aggiungerei nell'interfaccia Document:
```java
public interface Document {
    public void setTitle(String title);
    ...
}
```

e poi nel main puoi mettere un metodo che crea `n` documenti che hanno come titotolo `1`, `2`, `3`, ... .
In questo caso questo metodo deve fare la stessa cosa sia che si tratti di un PDF che di un file di testo, però senza il **Factory Method** dovresti costruire due volte lo stesso metodo con l'unica differenza che usa due costruttori diversi.
Il metodo sarebbe
```java
private Document[] createNDocumentsAndSetTitles(int n, DocumentFactory factory) {
    Document[] documents = new Documents[n];
    for (int i = 0; i < n; i++) {
        documents[i] = factory.createDocument();
        documents[i].setTitle(String.valueOf(i));
    }

    return documents;
}
```

E poi nel main giusto richiami due volte il metodo con le due factory